plugins/modules/cloud/misc/proxmox_disk.py

```diff
@@ -0,0 +1,697 @@
+
+    # Ensure VM id exists and retrieve its config
+    vm = None
+    vm_config = None
+    try:
+        vm = proxmox.get_vm(vmid)
+        vm_config = proxmox.proxmox_api.nodes(vm['node']).qemu(vmid).config.get()
+    except Exception as e:
+        proxmox.module.fail_json(msg='Getting information for VM %s failed with exception: %s' % (vmid, str(e)))
+
+    if state == 'present':
+        try:
+            proxmox.create_disk(disk, vmid, vm, vm_config)
+            module.exit_json(changed=True, vmid=vmid, msg="Disk %s created in VM %s" % (disk, vmid))
+        except Exception as e:
+            module.fail_json(vmid=vmid, msg='Unable to create disk %s in VM %s: %s' % (disk, vmid, str(e)))
+
+    # Do not try to perform actions on missing disk
+    if disk not in vm_config and state in ['absent', 'updated', 'grown', 'detached', 'moved']:
+        module.fail_json(vmid=vmid, msg='Unable to process missing disk %s in VM %s' % (disk, vmid))
+
+    elif state == 'absent':
+        try:
+            proxmox.proxmox_api.nodes(vm['node']).qemu(vmid).unlink.put(vmid=vmid, idlist=disk, force=1)
+            module.exit_json(changed=True, vmid=vmid, msg="Disk %s removed from VM %s" % (disk, vmid))
+        except Exception as e:
+            module.fail_json(vmid=vmid, msg='Unable to remove disk %s from VM %s: %s' % (disk, vmid, str(e)))
+
+    elif state == 'updated':
+        try:
+            proxmox.update_disk(disk, vmid, vm, vm_config)
+            module.exit_json(changed=True, vmid=vmid, msg="Disk %s updated in VM %s" % (disk, vmid))
+        except Exception as e:
+            module.fail_json(msg="Failed to updated disk %s in VM %s with exception: %s" % (disk, vmid, str(e)))
+
+    elif state == 'detached':
+        try:
+            if disk_bus == 'unused':
+                module.exit_json(changed=False, vmid=vmid, msg='Disk %s already detached in VM %s' % (disk, vmid))
+            proxmox.proxmox_api.nodes(vm['node']).qemu(vmid).unlink.put(vmid=vmid, idlist=disk, force=0)
+            module.exit_json(changed=True, vmid=vmid, msg="Disk %s detached from VM %s" % (disk, vmid))
+        except Exception as e:
+            module.fail_json(msg="Failed to detach disk %s from VM %s with exception: %s" % (disk, vmid, str(e)))
+
+    elif state == 'moved':
+        try:
+            proxmox.move_disk(disk, vmid, vm, vm_config)
+            disk_config = disk_conf_str_to_dict(vm_config[disk])
+            module.exit_json(
+                changed=True,
+                vmid=vmid,
+                storage=disk_config["storage_name"],
+                msg="Disk %s moved from VM %s storage %s" % (disk, vmid, disk_config["storage_name"]))
+        except Exception as e:
+            module.fail_json(msg="Failed to move disk %s in VM %s with exception: %s" % (disk, vmid, str(e)))
+
+    elif state == 'grown':
+        try:
+            size = module.params['size']
+            if not match(r'^\+?\d+(\.\d+)?[KMGT]?$', size):
```

Comment:
suggestion: you might want to allow an optional space between the number and the unit

Reply:
I would but API have strict format in size options.